### PR TITLE
KAFKA-8648: Throw exception if bad --property sent to ConsoleConsumer

### PR DESCRIPTION
**Pull request by astubbs**
Doesn’t check ConsoleProducer
Doesn’t check properties sent to key/value deserialisers

Includes unit test, also adds a test for the Custom Value Deserializer option. Also includes a more reliable way to test for system exit, if the code under test might throw an illegal argument exception.

For example, if you try to use

`./bin/kafka-console-consumer --topic c3test --bootstrap-server $BROKER:9092 --property interceptor.classes=io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor`

It will fail saying: 
`Unrecognised arguments ({interceptor.classes=io.confluent.monitoring.clients.interceptor.MonitoringConsumerInterceptor}) passed to kafka.tools.DefaultMessageFormatter.`

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by astubbs** [comment]
Retest this please

**Comment by astubbs** [comment]
I took a go at renaming the property, and deprecating the old property name

**Comment by astubbs** [comment]
Please test again

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala [hunk 1]
```diff
@@ -466,36 +482,62 @@ class DefaultMessageFormatter extends MessageFormatter {
...
-    // Note that `toString` will be called on the instance returned by `Deserializer.deserialize`
-    if (props.containsKey("value.deserializer")) {
-      valueDeserializer = Some(Class.forName(props.getProperty("value.deserializer")).getDeclaredConstructor()
-        .newInstance().asInstanceOf[Deserializer[_]])
-      valueDeserializer.get.configure(propertiesWithKeyPrefixStripped("value.deserializer.", props).asScala.asJava, false)
+
+    // https://github.com/scala/bug/issues/10418
+    props.putAll(propsOriginal: java.util.Map[_, _])
+
+    process("print.timestamp", (value: String) => printTimestamp = value.trim.equalsIgnoreCase("true"))
```
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala:498]: This is a pretty neat refactor :) Thanks

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala [hunk 2]
```diff
@@ -466,36 +482,62 @@ class DefaultMessageFormatter extends MessageFormatter {
...
+    keyDeserProps.asScala.foreach { case (key, value) =>
+      props.remove(keyDeserializerKey + '.' + key)
+    }
+    valDeserProps.asScala.foreach { case (key, value) =>
+      props.remove(valueDeserializerKey + '.' + key)
+    }
+
+    if (!props.isEmpty) {
+      // props contained an unrecognised property
+      throw new IllegalArgumentException(s"Unrecognised arguments (${props}) passed to ${this.getClass.getName}.")
```
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala:530]: Should we also add these checks to the other formatter classes - `LoggingMessageFormatter` and `ChecksumMessageFormatter`?

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala [hunk 3]
```diff
@@ -466,36 +482,62 @@ class DefaultMessageFormatter extends MessageFormatter {
...
     }
   }
 
+
+
```
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala:535]: nit: extra new lines

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala [hunk 4]
```diff
@@ -466,36 +482,62 @@ class DefaultMessageFormatter extends MessageFormatter {
...
   private def propertiesWithKeyPrefixStripped(prefix: String, props: Properties): Properties = {
     val newProps = new Properties()
     props.asScala.foreach { case (key, value) =>
       if (key.startsWith(prefix) && key.length > prefix.length)
-        newProps.put(key.substring(prefix.length), value)
+        newProps.put(key.substring(prefix.length + 1), value)
```
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala:540]: Why was this needed? Seems like you also had to change the tests for this to pass?

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala [hunk 5]
```diff
@@ -310,7 +322,11 @@ object ConsoleConsumer extends Logging {
       formatterArgs.setProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, valueDeserializer)
     }
 
-    formatter.init(formatterArgs)
+    try {
+      formatter.init(formatterArgs)
+    } catch {
+      case e:IllegalArgumentException => CommandLineUtils.printUsageAndDie(parser, e.getMessage)
```
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala:328]: nit: space after `e:`

File: core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala [hunk 6]
```diff
@@ -495,4 +514,76 @@ class ConsoleConsumerTest {
...
+    * There is no test for the default key and value deserializer properties
+    */
+  @Test(expected = classOf[ShouldExitException])
+  def shouldExitOnUnrecognisedProperty() {
+    val expectedMessage = Some("Unrecognised arguments ({not=ok, bad.property=not.good}) passed to kafka.tools.DefaultMessageFormatter.")
+    Exit.setExitProcedure((_, message) => {
+      assertEquals("Console consumer should exit with help information with bad properties listed", expectedMessage, message)
+      throw new ShouldExitException() // Code should exit with help information
+    }
+    )
```
Comment (stanislavkozlovski) [inline core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:528]: nit: formatting, the parentheses could be placed on top to avoid an extra new line

File: core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala [hunk 7]
```diff
@@ -495,4 +514,76 @@ class ConsoleConsumerTest {
...
   }
+
+  /**
+    * There is no test for the default key and value deserializer properties
+    */
+  @Test(expected = classOf[ShouldExitException])
+  def shouldExitOnUnrecognisedProperty() {
+    val expectedMessage = Some("Unrecognised arguments ({not=ok, bad.property=not.good}) passed to kafka.tools.DefaultMessageFormatter.")
+    Exit.setExitProcedure((_, message) => {
+      assertEquals("Console consumer should exit with help information with bad properties listed", expectedMessage, message)
```
Comment (stanislavkozlovski) [inline core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:525]: It's more brittle to assert the full error message. Perhaps we can assert that something smaller is in the message and still get the same assertion (that the properties are listed). e.g assert it contains `not=ok` or `Unrecognised`

File: core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala [hunk 8]
```diff
@@ -495,4 +514,76 @@ class ConsoleConsumerTest {
...
+
+    val args: Array[String] = Array(
+      "--bootstrap-server", "localhost:9092",
+      "--topic", "test",
+      "--property", "print.key=true",
+      "--property", "bad.property=not.good",
+      "--property", "not=ok")
+
+    try {
+      try {
```
Comment (stanislavkozlovski) [inline core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:538]: Any reason to have two `try`s rather than one? The finally block should always get executed, even if the `catch` raises an exception itself, I think

File: core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala [hunk 9]
```diff
@@ -495,4 +514,76 @@ class ConsoleConsumerTest {
...
+    assertTrue(config.formatter.isInstanceOf[DefaultMessageFormatter])
+    val formatter = config.formatter.asInstanceOf[DefaultMessageFormatter]
+    assertTrue("print.key setting picked up properly from new property name", formatter.printKey)
+  }
+
+  @Test(expected = classOf[ShouldExitException])
+  def shouldExitOnDeprecatedPropertyMix() {
+    val expectedMessage = Some("[property] is deprecated. It will be removed from future releases. Please use [formatter-property] instead. You cannot use both.")
+    Exit.setExitProcedure((_, message) => {
+      assertEquals("Console consumer should exit with help information about deprecated option", expectedMessage, message)
```
Comment (stanislavkozlovski) [inline core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:567]: ditto on the assertion and formatting

File: core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala [hunk 10]
```diff
@@ -495,4 +514,76 @@ class ConsoleConsumerTest {
       Exit.resetExitProcedure()
     }
   }
+
+  /**
+    * There is no test for the default key and value deserializer properties
```
Comment (stanislavkozlovski) [inline core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:519]: I don't fully understand this comment. can we not add one?

**Review by stanislavkozlovski — COMMENTED** [review]
Thanks for the PR @astubbs ! I made a quick pass over the code and left some small comments.

I agree with the intention of the JIRA but unfortunately I believe we might need a small KIP to get this change in. I also think it would be good to raise the desired behavior with the wider community such that we can converge on how we want our CLI tools to handle unknown arguments.

In this specific case, I think it is very clear that the `--property` flag is confusing and should be deprecated.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala [hunk 11]
```diff
@@ -466,36 +482,62 @@ class DefaultMessageFormatter extends MessageFormatter {
...
+      }
     }
-    // Note that `toString` will be called on the instance returned by `Deserializer.deserialize`
-    if (props.containsKey("value.deserializer")) {
-      valueDeserializer = Some(Class.forName(props.getProperty("value.deserializer")).getDeclaredConstructor()
-        .newInstance().asInstanceOf[Deserializer[_]])
-      valueDeserializer.get.configure(propertiesWithKeyPrefixStripped("value.deserializer.", props).asScala.asJava, false)
+
+    // https://github.com/scala/bug/issues/10418
+    props.putAll(propsOriginal: java.util.Map[_, _])
```
Comment (stanislavkozlovski) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala:496]: This is failing to build in Java8/Scala2.11 with:
```
19:13:47 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/core/src/main/scala/kafka/tools/ConsoleConsumer.scala:496: type mismatch;
19:13:47  found   : java.util.Map[_$3,_$4] where type _$4, type _$3
19:13:47  required: java.util.Map[_ <: Object, _ <: Object]
19:13:47     props.putAll(propsOriginal: java.util.Map[_, _])
```

**Comment by astubbs** [comment]
Thanks for the review @stanislavkozlovski ! Good comments, will process them and update - will be in touch.
